lib/concurrent/atomic/thread_local_var.rb

```diff
@@ -115,6 +115,45 @@ def set(value)
     end
   end
 
+  require 'monitor'
+  class AlexThreadLocalVar < AbstractThreadLocalVar
+
+    @@free = []
+    @@next = 0
+    @@lock = Monitor.new
+
+    protected
+
+    # PROBLEM: stale values could pass from a GC'd TLV to a newly allocated one!!
+
+    def self.finalizer(index)
+      lambda do
+        @@lock.synchronize do
+          @@free.push(index)
+        end
+      end
+    end
+
+    def allocate_storage
+      @index = @@lock.synchronize do
+        @@free.pop || begin
+          result = @@next
+          @@next += 1
+          result
+        end
+      end
+      ObjectSpace.define_finalizer(self, RubyThreadLocalVar.finalizer(@index))
```

Comment:
`ObjectSpace#each_object` with a non-`Module` subclass is disabled by default on JRuby, but other `ObjectSpace` methods, including `#define_finalizer` work fine.